### PR TITLE
Feature: Protocol TCP SYN timeout (init_timeout)

### PR DESCRIPTION
**Pull request by mpenning**
## 1. Problem statement

Legacy Exscript TCP timeout behavior is rather annoying... as of Exscript 2.1.440 there is no good way to control how long Exscript waits while trying to connect; by default it's at least 30 seconds (tested on Debian linux 7.1, kernel 3.2.0-4-amd64, Python 2.7.3, Exscript 2.1.440)...

    >>> from Exscript.protocols import SSH2
    >>> conn = SSH2() 
    >>> conn.connect('192.0.2.45')   # 192.0.2.45 doesn't exist

    ## Now wait 30 seconds...

    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "Exscript/protocols/Protocol.py", line 572, in connect
        return self._connect_hook(self.host, port, self.init_timeout)
      File "Exscript/protocols/Telnet.py", line 48, in _connect_hook
        receive_callback = self._telnetlib_received)
      File "Exscript/protocols/telnetlib.py", line 193, in __init__
        self.open(host, port)
      File "Exscript/protocols/telnetlib.py", line 223, in open
        raise socket.error, msg
    socket.timeout: timed out

We need a way to specify a socket-level TCP SYN timeout.  I created the ``init_timeout`` keyword in ``Exscript.protocols.Telnet`` and ``Exscript.protocols.SSH2``,  since ``Exscript.protocols.Protocol.timeout`` is already used...

## 2. Example SSH2 usage with the new ``init_timeout`` keyword:

    >>> from Exscript.protocols import SSH2
    >>> conn = SSH2(init_timeout=3)
    >>> conn.connect('192.0.2.45')   # 192.0.2.45 doesn't exist

    ## Now we only have to wait 3 seconds for ``socket.timeout``

## 3. Example SSH2 usage with the new ``init_timeout`` keyword:

    >>> from Exscript.protocols import Telnet
    >>> conn = Telnet(init_timeout=3)
    >>> conn.connect('192.0.2.45')   # 192.0.2.45 doesn't exist

    ## Now we only have to wait 3 seconds for ``socket.timeout``

## 4. ``make tests`` for this PR
    .............................................................................................................................................................................    ...............................................................................................................
    ----------------------------------------------------------------------
    Ran 500 tests in 198.295s

    OK

### CODE REVIEW

**Comment by egroeper** [comment]
Thanks for your pull request! It looks interesting, but I need to have a deeper look at it before I can merge.

File: src/Exscript/protocols/Protocol.py [hunk 1]
```diff
@@ -547,13 +550,13 @@ def get_timeout(self):
...
 
-    def _connect_hook(self, host, port):
+    def _connect_hook(self, host, port, init_timeout):
         """
         Should be overwritten.
         """
         raise NotImplementedError()
 
-    def connect(self, hostname = None, port = None):
+    def connect(self, hostname = None, port = None, init_timeout = None):
```
Comment (egroeper) [inline src/Exscript/protocols/Protocol.py]: You add the parameter init_timeout to the connect method, but it is never used.

**Comment by mpenning** [comment]
I updated the repo with [`commit fbd476f4ff`](https://github.com/mpenning/exscript/commit/fbd476f4ffb17e02fc6b88a88447c8f5bccf6d31)... please let me know if there is anything else you'd like to change.

`make tests` still passes...


    ...................................................................................................................................................................
    .........................................................................................................................

    ----------------------------------------------------------------------

    Ran 500 tests in 199.569s


    OK
    (py27_test)[mpenning@tsunami exscript]$

File: src/Exscript/protocols/Protocol.py [hunk 2]
```diff
@@ -561,10 +564,12 @@ def connect(self, hostname = None, port = None):
         @param hostname: The remote host or IP address.
         @type  port: int
         @param port: The remote TCP port number.
+        @type  init_timeout: int
```
Comment (egroeper) [inline src/Exscript/protocols/Protocol.py]: You don't expose that as a parameter, but you document it. I would remove the docstrings (and don't expose it).

File: src/Exscript/protocols/SSH2.py [hunk 3]
```diff
@@ -255,7 +255,7 @@ def _paramiko_shell(self):
             self._dbg(1, 'Failed to open shell.')
             raise LoginFailure('Failed to open shell: ' + str(e))
 
-    def _connect_hook(self, hostname, port):
+    def _connect_hook(self, hostname, port, init_timeout):
```
Comment (egroeper) [inline src/Exscript/protocols/SSH2.py]: You added an unused parameter. Probably to be compatible with Telnet implementation. But I really don't like that.
If it's there, it should do something. I would remove it anyway (see general pull request comment).

File: src/Exscript/protocols/Telnet.py [hunk 4]
```diff
@@ -36,11 +36,12 @@ def _telnetlib_received(self, data):
         self._receive_cb(data)
         self.buffer.append(data)
 
-    def _connect_hook(self, hostname, port):
+    def _connect_hook(self, hostname, port, init_timeout):
         assert self.tn is None
         rows, cols = get_terminal_size()
         self.tn = telnetlib.Telnet(hostname,
                                    port or 23,
+                                   init_timeout     = init_timeout,
```
Comment (egroeper) [inline src/Exscript/protocols/Telnet.py]: Please just use self.init_timeout and don't add init_timeout as parameter to  _connect_hook.

**Comment by egroeper** [comment]
Sorry for the delay. I think your change will improve exscript, but the current implementation contains some errors.
Looking at the code, I would suggest you only add init_timeout as constructor parameter to Protocol (and subclasses of course), but don't change _connect_hook or connect, as I don't think it's a parameter, that really changes per host and changing _connect_hook further complicates things.

Additionally you should implement set_init_timeout() and get_init_timeout(). That way it's easy to adjust the setting, if needed.

**Comment by mpenning** [comment]
@egroeper, renaming the parameter as ``connect_timeout`` is fine with me.  Regarding unit tests, a long-term desire I have is building something similar to [MockSSH](https://github.com/ncouture/MockSSH) which supports these kind of features; gevent looks promising so far.

**Comment by egroeper** [comment]
@mpenning: You mean MockSSH with socket magic?
@knipknap already wrote unit tests using SSH and telnet servers (src/Exscript/servers). This should work fine for most test cases, but I couldn't extend it to test the connect_timeout (see comment above).
I'm not sure how gevent could help here. Perhaps it could help speeding up testing.

**Pull request merged**